packages/-ember-data/tests/helpers/qunit-asserts/assert-deprecation.ts

```diff
@@ -0,0 +1,168 @@
+import QUnit from 'qunit';
+import { registerDeprecationHandler } from '@ember/debug';
+import { checkMatcher } from './check-matcher';
+import RSVP from 'rsvp';
+
+let HAS_REGISTERED = false;
+let DEPRECATIONS_FOR_TEST: FoundDeprecation[];
+let HANDLED_DEPRECATIONS_FOR_TEST: FoundDeprecation[];
+
+interface DeprecationConfig {
+  id: string;
+  count?: number;
+  until: string;
+  message?: string | RegExp;
+  url?: string;
+}
+interface FoundDeprecation {
+  message: string;
+  options: {
+    id: string;
+    message?: string;
+    until: string;
+    url?: string;
+  };
+}
+
+interface AssertSomeResult {
+  result: boolean;
+  actual: { id: string; count: number };
+  expected: { id: string; count: number };
+  message: string;
+}
+interface AssertNoneResult {
+  result: boolean;
+  actual: FoundDeprecation[];
+  expected: FoundDeprecation[];
+  message: string;
+}
+
+/**
+ * Returns a qunit assert result object which passes if the given deprecation
+ * `id` was found *exactly* `count` times.
+ *
+ * Fails if not found or found more or less than `count`.
+ * Fails if `until` not specified
+ * Optionally fails if `until` has been passed.
+ */
+function expectDeprecation(config: DeprecationConfig): AssertSomeResult {
+  // TODO optionally throw if `until` is the current version or older than current version
+  let matchedDeprecations = DEPRECATIONS_FOR_TEST.filter(deprecation => {
+    if (!deprecation.options || !deprecation.options.id) {
+      throw new Error(`Expected deprecation to have an id, found: ${deprecation}`);
+    }
+    let isMatched = deprecation.options.id === config.id;
+    if (!isMatched && config.message) {
+      // TODO when we hit this we should throw an error in the near future
+      isMatched = checkMatcher(deprecation.message, config.message);
+    }
+    return isMatched;
+  });
+  DEPRECATIONS_FOR_TEST = DEPRECATIONS_FOR_TEST.filter(deprecation => {
+    matchedDeprecations.indexOf(deprecation) === -1;
+  });
+  HANDLED_DEPRECATIONS_FOR_TEST.push(...matchedDeprecations);
+
+  let expectedCount = typeof config.count === 'number' && config.count !== 0 ? config.count : 1;
+  let passed = matchedDeprecations.length === expectedCount;
+
+  return {
+    result: passed,
+    actual: { id: config.id, count: matchedDeprecations.length },
+    expected: { id: config.id, count: expectedCount },
+    message: `Expected ${expectedCount} deprecation${expectedCount === 1 ? '' : 's'} for ${config.id} during test, ${
+      passed ? expectedCount : 'but ' + matchedDeprecations.length
+    } deprecations were found.`,
+  };
+}
+
+function expectNoDeprecation(): AssertNoneResult {
```

Comment:
Lets call this method something different, its pretty confusing to have both `QUnit.assert.expectNoDeprecation` and this inline function with the same name but quite different.

I'm thinking something like `verifyNoDeprecationsIssued`...